### PR TITLE
timers: Add timer_spin_delay_{us,ns}

### PR DESCRIPTION
**Pull request by pcercuei**
These two functions can be used to busy-wait until a given timeout expires. They are similar to timer_spin_sleep(), except that they are much more precise and don't use a dedicated TIMER UNIT channel, which makes them inherently thread-safe.

### CODE REVIEW

File: kernel/arch/dreamcast/kernel/timer.c [hunk 1]
```diff
@@ -171,6 +171,22 @@ void timer_spin_sleep(int ms) {
     timer_stop(TMU1);
 }
 
+void timer_spin_sleep_ns(unsigned short ns) {
+    uint64_t timeout = timer_ns_gettime64() + ns;
```
Comment (gyrovorbis) [inline kernel/arch/dreamcast/kernel/timer.c:175]:
```suggestion
    const uint64_t timeout = timer_ns_gettime64() + ns;
```

File: kernel/arch/dreamcast/kernel/timer.c [hunk 2]
```diff
@@ -171,6 +171,22 @@ void timer_spin_sleep(int ms) {
...
+
+    /* Note that we don't actually care about the counter overflowing.
+       Nobody will run their Dreamcast straight for 585 years. */
+    while(timer_ns_gettime64() < timeout);
+}
+
+void timer_spin_sleep_us(unsigned short us) {
+    uint64_t timeout = timer_us_gettime64() + us;
```
Comment (gyrovorbis) [inline kernel/arch/dreamcast/kernel/timer.c:183]:
```suggestion
    const uint64_t timeout = timer_us_gettime64() + us;
```

File: kernel/arch/dreamcast/kernel/timer.c [hunk 3]
```diff
@@ -171,6 +171,22 @@ void timer_spin_sleep(int ms) {
     timer_stop(TMU1);
 }
 
+void timer_spin_sleep_ns(unsigned short ns) {
```
Comment (gyrovorbis) [inline kernel/arch/dreamcast/kernel/timer.c:174]:
```suggestion
void timer_spin_delay_ns(unsigned short ns) {
```

File: kernel/arch/dreamcast/kernel/timer.c [hunk 4]
```diff
@@ -171,6 +171,22 @@ void timer_spin_sleep(int ms) {
...
+    uint64_t timeout = timer_ns_gettime64() + ns;
+
+    /* Note that we don't actually care about the counter overflowing.
+       Nobody will run their Dreamcast straight for 585 years. */
+    while(timer_ns_gettime64() < timeout);
+}
+
+void timer_spin_sleep_us(unsigned short us) {
```
Comment (gyrovorbis) [inline kernel/arch/dreamcast/kernel/timer.c:182]:
```suggestion
void timer_spin_delay_us(unsigned short us) {
```

**Comment by gyrovorbis** [comment]
So I couldn't find my other SD card to format and use with either of KOS's examples that do much, but I was able to verify that there were no performance regressions with @DC-SWAT's nice performance test.

Before the changes:
Screenshot 2024-05-10 23-29-18

After the changes:
Screenshot 2024-05-10 23-34-24

**Comment by andressbarajas** [comment]
@zcrc I tested by pulling your branch and executing the mke2fs example to completion on 8gb sd card

File: kernel/arch/dreamcast/kernel/timer.c (hunk 1 above)
Comment (pcercuei) [reply kernel/arch/dreamcast/kernel/timer.c:175]: Ehrm, what's the point of using "const" here?

File: kernel/arch/dreamcast/kernel/timer.c (hunk 3 above)
Comment (pcercuei) [reply kernel/arch/dreamcast/kernel/timer.c]: Ouch. Stupid mistake, will fix.

**Comment by andressbarajas** [comment]
@pcercuei Copyrights to timer.h/c, CHANGELOG?

**Comment by pcercuei** [comment]
@andressbarajas done.

**Review by andressbarajas — APPROVED** [review]

File: kernel/arch/dreamcast/kernel/timer.c (hunk 1 above)
Comment (gyrovorbis) [reply kernel/arch/dreamcast/kernel/timer.c:175]: Nothing, really, the compiler is smart enough. I'm just anal. If you don't really care, then I guess I'll back off. lol.

**Comment by pcercuei** [comment]
Rebased on top of master.

**Comment by andressbarajas** [comment]
For whomever reviews this PR, there are no unresolved conversations and if everything looks good to you then feel free to merge.

**Comment by gyrovorbis** [comment]
Wait, didn't this PR also need to reimplement `timer_spin_sleep()` without taking up the `TMU` channel 1 in the same manner? I also think maybe we should deprecate `timer_spin_sleep()` (in name) and add the new function (that it calls into for now) as `timer_spin_delay_ms()`?

So `timer_spin_sleep()` is a legacy API thing that will be explicitly  marked `\deprecated` in the Doxygen, and will be an `inline static` function that calls into `timer_spin_delay_ms()` which will then use this same mechanism to implement the delay?

**Review by QuzarDC — APPROVED** [review]

**Review by gyrovorbis — APPROVED** [review]
I'm willing to merge this one myself... Forgive my audacity, but I'm figuring:

1) At this point, I'm super familiar with the TMUs and driver
2) We aren't breaking any existing API
3) The SCIF change I have tested personally
4) It's gating another draft PR I'm working on involving timers.
5) It's part of a big-picture scheme that frees up a TMU channel for POSIX timers eventually. :)

**Pull request merged**